xarray/core/dataset.py

```diff
@@ -2980,20 +2980,31 @@ def isel(
         coord_names = self._coord_names.copy()
 
         indexes, index_variables = isel_indexes(self.xindexes, indexers)
+        all_keys = set(indexers.keys())
 
         for name, var in self._variables.items():
             # preserve variable order
             if name in index_variables:
                 var = index_variables[name]
-            else:
-                var_indexers = {k: v for k, v in indexers.items() if k in var.dims}
-                if var_indexers:
+                dims.update(zip(var.dims, var.shape))
+            # Fastpath, skip all of this for variables with no dimensions
+            # Keep the result cached for future dictionary update
+            elif var_dims := var.dims:
+                # Large datasets with alot of metadata may have many scalars
+                # without any relevant dimensions for slicing.
+                # Pick those out quickly and avoid paying the cost below
+                # of resolving the var_indexers variables
+                var_indexer_keys = all_keys.intersection(var_dims)
+                if var_indexer_keys:
+                    var_indexers = {k: indexers[k] for k in var_indexer_keys}
                     var = var.isel(var_indexers)
                     if drop and var.ndim == 0 and name in coord_names:
                         coord_names.remove(name)
                         continue
+                    # Update after slicing
+                    var_dims = var.dims
+                dims.update(zip(var_dims, var.shape))
```

Comment:
I didn;t understand "update after slicing". Can you add a longer comment please?

Also i believe the `dims.update(...)` line can stay unmodified?

Reply:
Please see suggestion below regarding the comment.
```suggestion
                    # Update our reference to `var_dims` after the call to isel
                    var_dims = var.dims
                dims.update(zip(var_dims, var.shape))
```

Reply:
I think a little more detail would be nice but perhaps this is too clever.

IIUC you're trying to speed up a case where you're iterating over variables with the same dimensions one after the other.

Would it help your use-case to special case scalars with a `continue` instead? (assuming that is correct)